Implement event API in core module

jakobkmar opened the pull request:
### Event API

This implements an event API in the silk-core module.
Additionally it adds some basic first events.

*For reviewers:* the relevant API part is located here https://github.com/SilkMC/silk/blob/events/silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt

Please comment on this PR to suggest changes, since this API is not necessarily final yet!

### Goals

Provide an event API for the following situations:
- very lightweight event handling with a list of listeners saved in each event instance
- coroutine based event handling via flows, lightweight solution for async event handling

The use case for handlers to mutate and or cancel the event must be supported.
Reflection must be avoided.

### Usage:

**To declare an event**
```kt
// really simple, a list of listeners which will be invoked in place
val myEvent = Event.onlySyncImmutable<MyEvent>()
// the same but cancellable
val myEvent = Event.onlySync<MyEvent, EventScope.Cancellable>()

// can be listened to both synchronously and asynchronously via a flow
// there is a mutable scope for synchronous listeners, if they want to cancel etc
val myEvent = Event.syncAsync<MyEvent, EventScope.Cancellable>()

// the event scope is always immutable and listeners may be synchronous or asynchronous
val myEvent = Event.syncAsyncImmutable<MyEvent>()
```

**To make your event discoverable**, use an extension function:
```kt
// do not pollute the main events object, use your own namespace
val Events.MyModSlug get() = MyModSlugEvents
object MyModSlugEvents

// categorize events under your namespace if needed
val MyModSlugEvents.MySpecialEvents get() = MySpecialEvents

// define the events
object MySpecialEvents {
    val myEvent = Event.onlySync<MyEvent>()
}
```

**Now you can listen to the event:**
```kt
// simple synchronous listener
Events.MyModSlug.MySpecialEvents.myEvent.listen {

}

// the same with a different priority
Events.MyModSlug.MySpecialEvents.myEvent.listen(EventPriority.FIRST) {

}

// or collect the flow
// in scope
coroutineScope {
    Events.MyModSlug.MySpecialEvents.myEvent.collectInScope {
        
    }
}
// or bare bones (this never completes)
Events.MyModSlug.MySpecialEvents.myEvent.collect {
        
}
```
Note: silk itself does not use a separate namespace, to shorten the most commonly used events

**To invoke the event:**
from Kotlin:
```kt
MySpecialEvents.myEvent.invoke(MyEvent())
```
from Java (needed for Mixins):
```java
MySpecialEvents.INSTANCE.getMyEvent().invoke(MyEvent())
```

For mutable events, you can pass an event scope, see https://github.com/SilkMC/silk/blob/events/silk-core/src/main/java/net/silkmc/silk/core/mixin/server/MixinMinecraftServer.java#L59

F0Xde reviewed (commented):
Looks good, just a few questions mostly about mutation/cancellation:

- Are there any plans to support and enforce event cancellation via silk or is this something the user has to handle themselves by mutating and checking if the event is cancelled manually?
- Will it be possible to return a value to the invoker so that the cause can for example be cancelled?
- What are your thoughts on some kind of priority system? It seems to me that without such system the order in which listeners are called is mostly random (assuming for example all listeners are registered in the initialization function of their mods) which would make these features somewhat useless or at least very hard to use effectively as soon as multiple listeners want to work on the same event.

jakobkmar commented:
Thanks for the review!

> Are there any plans to support and enforce event cancellation via silk or is this something the user has to handle themselves by mutating and checking if the event is cancelled manually?

> Will it be possible to return a value to the invoker so that the cause can for example be cancelled?

Both are supported by mutable event classes which need to be handled synchronously. Some base interfaces, e.g. for cancellable events will be added to this PR.

> What are your thoughts on some kind of priority system? It seems to me that without such system the order in which listeners are called is mostly random (assuming for example all listeners are registered in the initialization function of their mods) which would make these features somewhat useless or at least very hard to use effectively as soon as multiple listeners want to work on the same event.

That's definitely something which must be supported, and basically the reason why this PR is still a draft. I still have to think about priority + dynamic registration + flows will look like.

jakobkmar commented on silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt:
```diff
@@ -0,0 +1,136 @@
+     * listeners and collectors.
+     */
+    open fun invoke(instance: T) {
+        synchronized(this) {
```
It's not ideal, but the API should always be threadsafe and simply removing this would mean that the user has to be extra careful, which is not ideal. So this synchronized block should stay there *until* there is a better solution for a thread safe basic callback list.

F0Xde replied:
Yep totally agree, we would need a better solution that still allows usage from any thread

jakobkmar commented on silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt:
```diff
@@ -0,0 +1,136 @@
+    protected open val syncDispatcher: CoroutineDispatcher
+        get() = mcCoroutineDispatcher
+
+    /**
+     * Listens to this event, and `emit`s to the [collector] if
+     * it is called.
+     * Launches [SharedFlow.collect] on the underlying [flow] in the current
+     * [CoroutineScope] (taken from context), using the dispatcher from the
+     * current scope.
+     */
+    context(CoroutineScope)
+    fun collectInScope(collector: FlowCollector<ImmutableType>): Job = launch {
+        flow.collect(collector)
+    }
+
+    /**
+     * Listens to this event, and `emit`s to the [collector] if
+     * it is called.
+     * Launches [SharedFlow.collect] on the underlying [flow] in the current
+     * [CoroutineScope] (taken from context), using the fitting synchronous
+     * Minecraft main thread dispatcher ([syncDispatcher]).
+     */
+    context(CoroutineScope)
+    fun collectInScopeSync(collector: FlowCollector<ImmutableType>): Job = launch(syncDispatcher) {
+        flow.collect(collector)
+    }
+
+    /**
+     * Calls [SharedFlow.collect] on the underlying [flow].
+     * This way the [collector] receives all events.
+     * Note that the thread and / or [CoroutineScope] you call this in
+     * decides on which thread you handle the events.
+     *
+     * This function never completes, read [the official collect docs][kotlinx.coroutines.flow.SharedFlow.collect]
+     * for more info.
+     */
+    suspend fun collect(collector: FlowCollector<ImmutableType>): Nothing {
+        flow.collect(collector)
+    }
```
I will add example code snippets to the docs.

jakobkmar replied:
Samples will be added later using the testmod and referenced / made available to Dokka via the `@sample` KDoc attribute.

jakobkmar commented on silk-commands/src/main/kotlin/net/silkmc/silk/commands/event/CommandEvents.kt:
```diff
@@ -0,0 +1,41 @@
+        val selection: CommandSelection,
+    )
+
+    class RegisterEventClient(
```
Not sure, do we want to specify that an event is client specific at the end of the name in any case (more consistent) or always at the best place for the given context (may be inconsistent).

F0Xde replied:
I think having `Event` at the end makes the most sense, but I'm not sure if we should put the `Client` right before `Event` or always at the start (`ClientRegisterEvent` in this case) or choose depending on context

l4zs replied:
I'd prefer the prefix, but either way it should be consistent. Maybe even adding Server prefix although thats up to prefence as there shouldn't be any confusion.

jakobkmar commented on silk-commands/src/main/kotlin/net/silkmc/silk/commands/event/CommandEvents.kt:
```diff
@@ -0,0 +1,41 @@
+ * Events related to commands.
+ */
+@Suppress("unused") // receiver is for namespacing only
+val Events.command get() = CommandEvents
```
My decision to decapitalize these values was simply a cosmetic one, so it is not set in stone yet, I just thought it looks better / cleaner when listening to an event.

jakobkmar replied:
Capitalized them now.

jakobkmar commented on silk-core/src/main/kotlin/net/silkmc/silk/core/event/ServerEvents.kt:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Called before the server is initialized.
+     */
+    val preStart = Event.bothImmutable<ServerEvent>()
```
A also think that `starting` looks cleaner, but `preStart` immediately tells the user what exactly it does, so that was more important for me personally.

F0Xde replied:
Makes sense, `starting` kinda sounds like it's async and the server is starting in parallel

l4zs commented on buildSrc/build.gradle.kts:
```diff
@@ -15,20 +15,20 @@ repositories {
+    implementation(pluginDep("com.modrinth.minotaur", "2.3.1"))
 
-    implementation(pluginDep("io.papermc.paperweight.userdev", "1.3.6"))
+    implementation(pluginDep("io.papermc.paperweight.userdev", "1.3.7"))
```
there's 1.3.8 already iirc

l4zs commented on silk-core/src/main/java/net/silkmc/silk/core/mixin/server/MixinMinecraftServer.java:
```diff
@@ -18,8 +18,39 @@ public class MixinMinecraftServer {
+        at = @At(
+            value = "INVOKE",
+            target = "Lnet/minecraft/server/MinecraftServer;initServer()Z",
+            shift = At.Shift.BEFORE
```
shouldn't this be `AFTER`?

l4zs commented on silk-core/src/main/kotlin/net/silkmc/silk/core/event/ServerEvents.kt:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Called after the server has successfully started.
+     */
+    val started = Event.bothImmutable<ServerEvent>()
```
I kinda think it'd be more consistent to have this named `postStart`

jakobkmar replied:
I decided to go with your suggestion and named them `postStart` and `postStop`.

l4zs reviewed (commented):
Looks good! Haven't got much to add. For the priorities I think the way Bukkit did it works pretty well if you follow their [guide](https://bukkit.fandom.com/wiki/Event_API_Reference#Event_Priorities) on what priority to use for your listeners.

jakobkmar commented on silk-core/src/main/java/net/silkmc/silk/core/mixin/server/MixinMinecraftServer.java:
```diff
@@ -18,8 +18,39 @@ public class MixinMinecraftServer {
             shift = At.Shift.BEFORE
```
If it is not harmful I think this makes it easier for readers to understand what this mixin does.

jakobkmar commented:
Added mutable event scopes to support cancellable events. Applied all suggestions and introduced [event priority](https://github.com/SilkMC/silk/blob/events/silk-core/src/main/kotlin/net/silkmc/silk/core/event/EventPriority.kt) to synchronous listeners.
Additionally, I added [a test to cover the event system](https://github.com/SilkMC/silk/blob/events/silk-core/src/test/kotlin/net/silkmc/silk/core/test/EventTest.kt).

A re-review / approval would be appreciated!

F0Xde reviewed (commented):
Great work, thanks!

There are a few optimizations and additional features that I have been thinking about, but I'd say we can postpone these to another PR.

jakobkmar commented on silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt:
```diff
@@ -0,0 +1,173 @@
+        synchronized(this) {
+            listenersByPriority[priority.ordinal] = listenersByPriority
+                .getOrElse(priority.ordinal) { ArrayList() }
+                .apply { add(callback) }
```
Good point, it is left over from before the commit where I initialized the list instantly. Changed with d356ba64

l4zs reviewed (approved):
Looks good to me, thanks for the great work done here!

jakobkmar commented on silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt:
```diff
@@ -0,0 +1,173 @@
+     * Invokes this event. Calling this function will trigger all
+     * listeners and collectors.
+     */
+    open fun invoke(instance: T, scope: S) {
```
I introduced a `scopeSupplier` to the Event constructor, so a default scope is always available and [a simple invoke function with just one parameter](https://github.com/SilkMC/silk/blob/dadfe5cb5101455ebc98e7249d7a14ce931537d1/silk-core/src/main/kotlin/net/silkmc/silk/core/event/Event.kt#L92) can be provided, which also returns the scope.

jakobkmar commented:
Thanks for all the feedback, really helped a lot to iterate over the API and improve its design!
